graph/generate.go

```diff
@@ -0,0 +1,254 @@
+/*
+Package graph provides core lineage graph services. It defines the graph datatype and provides the "New"
+function for creating one.
+*/
+package graph
+
+import (
+	"fmt"
+
+	mbp "github.com/GoogleCloudPlatform/healthcare-data-harmonization/mapping_engine/proto"
+	proto "github.com/golang/protobuf/proto"
+)
+
+var idFactory IDfactory = &autoIncFactory{currentID: 0}
+
+// setIDfactory sets the interface used for ID generation
+func setIDfactory(factory IDfactory) {
+	idFactory = factory
+}
+
+// msgContext is a wrapper around a proto message storing additional context, like parent node.
+type msgContext struct {
+	msg        proto.Message
+	descendant Node
+}
+
+type stack []msgContext
```

Comment:
nit: I am not a big fan of the APIs designed here: we are essentially exposing the implementation details of the stack data structure, which is OK since it is internal to this graph package. However, we have to update the method receiver in `reversePush` because of this, which can be a source of surprises down the road. I suggest we use a temporary variable in this method and return that instead.

Reply:
I changed stack to be a struct containing a list; the stack remains the same while the list it contains changes on push/pop.